mods/tnt/init.lua

```diff
@@ -544,9 +550,18 @@ function tnt.register_tnt(def)
 			is_ground_content = false,
 			groups = {dig_immediate = 2, mesecon = 2, tnt = 1, flammable = 5},
 			sounds = default.node_sound_wood_defaults(),
+			after_place_node = function(pos, placer)
+				if placer:is_player() then
+					local meta = minetest.get_meta(pos)
+					meta:set_string("owner", placer:get_player_name())
+				end
+			end,
 			on_punch = function(pos, node, puncher)
 				if puncher:get_wielded_item():get_name() == "default:torch" then
+					local meta = minetest.get_meta(pos)
+					local owner = meta:get_string("owner")
 					minetest.set_node(pos, {name = name .. "_burning"})
+					meta:set_string("owner", owner)
```

Comment:
While making the suggested change, I found out that the `on_construct()` call back is triggered by `set_node()` but not by `swap_node()`.  Missing that callback would have prevented the sound from playing, the node from falling, and the node timer from starting, which causes the `tnt:tnt_burning` to never explode.  To resolve these issues, I've added a manual call to `on_construct()` after each `swap_node()`.